src/rt/region/region_rc.h

```diff
@@ -181,16 +196,260 @@ namespace verona::rt
+     *  original refcount restored and is marked green.
+     *
+     *  3. o's subgraph is re-traced a final time, and any remaining red objects
+     *  are deallocated.
+     **/
+    static void gc_cycles(Alloc& alloc, Object* o)
+    {
+      RegionRc* reg = get(o);
+      ObjectStack jump_stack(alloc);
+      while (!reg->lins_stack.empty())
+      {
+        auto p = reg->lins_stack.pop(alloc);
+
+        if (p->get_rc_colour() == RcColour::BLACK)
+        {
+          mark_red(alloc, p, o, jump_stack);
+          scan(alloc, p, o, jump_stack);
+        }
+      }
+    }
+
   private:
-    inline static bool decref_inner(Object* o)
+    /**
+     * Mark an object as red (i.e. dead). This is recursive over o's fields
+     * where the field points to an objects which is also an rc candidate.
+     *
+     * mark_red traces the entire subgraph of o, decrefing on each edge and
+     * marking each subobject as red.
+     *
+     * Note that objects are marked red indiscriminately. In other words, an
+     * object `o` will be marked red even if it has a reference count greater
+     * than 0 (indicating a potential external reference is keeping it alive).
+     *
+     * This eagerness is based on the observation by Lins that objects in a
+     * cycle can have a non-zero count because they are referenced by other
+     * objects in the same subgraph which are yet to be processed. This means it
+     * is usually a good idea to eagerly mark them red but add them to a "jump
+     * stack" for their liveness to be confirmed later on. This is a performance
+     * optimisation that can result in fewer passes over the graph.
+     **/
+    static void
+    mark_red(Alloc& alloc, Object* o, Object* in, ObjectStack& jump_stack)
     {
-      RefCount* rc = o->get_ref_count();
-      if (rc->metadata == 1)
+      if (!o->is_rc_candidate())
       {
-        return true;
+        return;
+      }
+
+      if (o->get_rc_colour() != RcColour::RED)
+      {
+        o->set_rc_colour(RcColour::RED);
+      }
+
+      ObjectStack dfs(alloc);
+      o->trace(dfs);
+
+      while (!dfs.empty())
+      {
+        Object* f = dfs.pop();
+        if (!o->is_rc_candidate())
+        {
+          continue;
+        }
+
+        if (f->get_class() == RegionMD::ISO && f != in)
+        {
+          // If f is the entry point to a subregion we don't want to trace
+          // into it: the colour of `f` is enough to determine the fate of
+          // the entire subregion.
+          continue;
+        }
+
+        decref_inner(f);
+
+        if (f->get_rc_colour() == RcColour::RED)
+        {
+          // base case to stop us infinitely iterating a cycle.
+          continue;
+        }
+
+        if (f->get_ref_count()->metadata > 0)
+        {
+          // f might still be live, so add it to the jump stack to confirm
+          // later on.
+          //
+          // Checking if `f` already exists in the stack is expensive and
```

Comment:
That's correct at the point when we add things to the jump stack. But what this comment is saying is that we may add duplicates because an object may have more than one incoming edge.

What this is saying is that when we process objects in the jump stack, we don't actually need to worry about duplicates though, because if an item was restored green, when we pop it again we can just ignore it.

Reply:
But it will be `red` after the first addition to the jump stack, so you won't get duplicates.